starter-core/src/main/java/io/micronaut/starter/build/dependencies/DependencyContext.java

```diff
@@ -31,4 +44,70 @@ default void addDependency(@NonNull Dependency.Builder dependency) {
+                        return IS_COMPILE_API_OR_RUNTIME.test(d) || d.getScope().getPhases().contains(Phase.ANNOTATION_PROCESSING);
+                    }
+                    return IS_COMPILE_API_OR_RUNTIME.test(d);
+
+                })
+                .toList();
+        List<Dependency> dependenciesInMainClasspathWithoutDuplicates = filterDuplicates(dependenciesInMainClasspath);
+
+        List<Dependency> dependenciesInTestClasspath = dependencies.stream()
+                .filter(d -> d.getScope().getSource() == Source.TEST && IS_COMPILE_API_OR_RUNTIME.test(d))
+                .toList();
+
+        List<Dependency> dependenciesInTestClasspathWithoutDuplicates = filterDuplicates(dependenciesInTestClasspath);
+
+        dependenciesInTestClasspathWithoutDuplicates.removeIf(testDep -> {
+            MavenCoordinate test = new MavenCoordinate(testDep.getGroupId(), testDep.getArtifactId(), testDep.getVersion());
+            return dependenciesInMainClasspathWithoutDuplicates.stream()
+                    .filter(mainDep -> (mainDep.getScope().getPhases().contains(RUNTIME) && mainDep.getScope().getPhases().contains(COMPILATION)))
+                    .anyMatch(mainDep -> {
+                        MavenCoordinate main = new MavenCoordinate(mainDep.getGroupId(), mainDep.getArtifactId(), mainDep.getVersion());
+                        return main.equals(test);
+                    });
+        });
+        List<Dependency> result = new ArrayList<>(dependenciesNotInMainOrTestClasspath);
+        result.addAll(dependenciesInMainClasspathWithoutDuplicates);
+        result.addAll(dependenciesInTestClasspathWithoutDuplicates);
+        return result.stream().sorted(Dependency.COMPARATOR).toList();
+    }
+
+    private static List<Dependency> filterDuplicates(List<Dependency> dependencies) {
+        List<Dependency> dependenciesWithoutDuplicates = new ArrayList<>();
+        Map<MavenCoordinate, Scope> dependenciesWithScope = new HashMap<>();
+        for (Dependency dep : dependencies.stream().sorted(Dependency.COMPARATOR).toList()) {
+            MavenCoordinate coordinate = new MavenCoordinate(dep.getGroupId(), dep.getArtifactId(), dep.getVersion());
+            if (dependenciesWithScope.containsKey(coordinate)) {
+                if (dependenciesWithScope.get(coordinate).getOrder() < dep.getOrder()) {
+                    dependenciesWithScope.remove(coordinate);
+                    dependenciesWithoutDuplicates.removeIf(f -> new MavenCoordinate(f.getGroupId(), f.getArtifactId(), f.getVersion()).equals(coordinate));
+                    dependenciesWithScope.put(coordinate, dep.getScope());
+                    dependenciesWithoutDuplicates.add(dep);
+                }
+            } else {
+                dependenciesWithScope.put(coordinate, dep.getScope());
+                dependenciesWithoutDuplicates.add(dep);
+            }
+        }
+        return dependenciesWithoutDuplicates;
```

Comment:
```suggestion
        return new ArrayList<>(dependencies.stream().collect(
                Collectors.toMap(
                        d -> new MavenCoordinate(d.getGroupId(), d.getArtifactId(), d.getVersion()),
                        Function.identity(),
                        BinaryOperator.minBy(Comparator.comparing(d -> d.getScope().getOrder()))
                )
        ).values());
```

I *think* that's the same 🤔